### PR TITLE
Benchmarking in CI (SCP-2353)

### PR DESCRIPTION
**Pull request by gilligan**
**Summary**
Add a benchmarking script that will be executed by buildkite and a
GitHub action to trigger the buidkite pipeline running this script
through `/benchmark` comments on PRs

**Details**
The benchmark.yaml workflow below looks for `/benchmark` comments which works nicely *but* it should be noted this only works with this code in `master`. 

I did test and experiment using a fork and everything seemed to work nicely eventually but it's of course possible that something will still go wrong as we can't really test this fully before it's actually merged (annoying, yes).

The comments added by the bot aren't super pretty but i guess that is something we can change down the road once it's been established that this actually works reliably. Comments have to be posted as JSON content which is a bit annoying and I just used `jq` to escape the output generated by the benchmark comparison script.

------

Pre-submit checklist:
- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

File: .github/workflows/benchmark.yml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        if: steps.check.outputs.triggered == 'true'
+        id: comment-branch
+        with:
+          repo_token: ${{ secrets.GITHUB_TOKEN }}
+
+      #
+      # Trigger the buildkite pipeline IF the 'benchmark' command was found
+      #
+      - run: |
+         if [ -z "$BUILDKITE_API_ACCESS_TOKEN" ] ; then
```
[the remark this answers is not in the document]
Comment (michaelpj) [reply .github/workflows/benchmark.yml:37]: Ah, I see we template in some parameters. We could make them script parameters and then do `./trigger-pipeline.sh ${{templated_arg}}` or whatever.

File: ci-plutus-benchmark.sh [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+set -e
+
+if [ -z "$PR_NUMBER" ] ; then
+   echo "[ci-plutus-benchmark]: 'PR_NUMBER' is not set! Exiting"
+   exit 1
+fi
+echo "[ci-plutus-benchmark]: Processing benchmark comparison for PR $PR_NUMBER"
+
+echo "[ci-plutus-benchmark]: Updating cabal database ..."
```
Comment (michaelpj) [inline ci-plutus-benchmark.sh:27]: worth a note about why we need this

File: ci-plutus-benchmark.sh [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+cabal update
+
+echo "[ci-plutus-benchmark]: Running benchmark for PR branch ..."
+cabal bench plutus-benchmark:validation >bench-PR.log 2>&1
```
Comment (michaelpj) [inline ci-plutus-benchmark.sh:31]: Hmm, we should probably run this inside a `nix-shell` too!

File: ci-plutus-benchmark.sh [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+echo "[ci-plutus-benchmark]: Switching branches ..."
+git checkout "$(git merge-base HEAD master)"
+
+echo "[ci-plutus-benchmark]: Running benchmark for base branch ..."
+cabal bench plutus-benchmark:validation >bench-base.log 2>&1
```
Comment (michaelpj) [inline ci-plutus-benchmark.sh:37]: ditto

File: ci-plutus-benchmark.sh [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+echo "[ci-plutus-benchmark]: Comparing results ..."
+./plutus-benchmark/bench-compare bench-base.log bench-PR.log >bench-compare-result.log
+nix-shell -p jq --run "jq -Rs '.' bench-compare-result.log >bench-compare.json"
```
Comment (michaelpj) [inline ci-plutus-benchmark.sh:41]: worth a comment about what this is doing

**Review by michaelpj — COMMENTED** [review]
nice!

File: ci-plutus-benchmark.sh (hunk 3 above)
Comment (gilligan) [reply ci-plutus-benchmark.sh:31]: The whole script is being executed inside a nix-shell though:

```
steps:
  - command:
    - "nix-shell --run './ci-plutus-benchmark.sh'"   
    agents:
      queue: "plutus"
```

So this should be fine right?

File: scripts/trigger-buildkite-pipeline.sh [hunk 6]
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env bash
```
Comment (gilligan) [inline scripts/trigger-buildkite-pipeline.sh:1]: @michaelpj  now that I extracted it i agree that it's nicer to have this in a separate file/script :+1:

File: scripts/trigger-buildkite-pipeline.sh [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+if [ -z "$BUILDKITE_API_ACCESS_TOKEN" ] ; then
+   echo "[trigger-buildkite-pipeline]: 'BUILDKITE_API_ACCESS_TOKEN' is not set!"
+   exit 1
+fi
+
+BRANCH="$1"
+PR="$2"
+
+echo "[trigger-buildkite-pipeline]: Triggering build for $PR_BRANCH"
```
Comment (michaelpj) [inline scripts/trigger-buildkite-pipeline.sh:20]:
```suggestion
echo "[trigger-buildkite-pipeline]: Triggering build for $BRANCH"
```

**Review by michaelpj — APPROVED** [review]

**Pull request merged**